### PR TITLE
Remove `environs`

### LINKED ISSUE #2222: Remove environs
See Dave's comment in this [PR](https://github.com/opensafely-core/opencodelists/pull/2220#issuecomment-2531829119).

Airlock specifically didn't use environs because its transitive dependencies seemed pretty heavyweight for the value it gives. We're only using it in settings.py, so it seems unlikely we're getting the full value from it and that we can easily remove it.

### PR DESCRIPTION
**Pull request by Jongmassey**
fixes #2222

### CODE REVIEW

File: opencodelists/settings.py [hunk 1]
```diff
@@ -30,29 +30,58 @@
...
+have been updated since `.env` was first created).
+
+If you are seeing this error in production then you haven't configured
+things properly.
+"""
+
+
+def get_env_var(name, **kwargs):
+    try:
+        env = os.environ[name]
```
Comment (mikerkelly) [inline opencodelists/settings.py:52]: I'm not sure about `env` as a variable name. `env_var_value`?

File: opencodelists/settings.py [hunk 2]
```diff
@@ -30,29 +30,58 @@
...
+        if "default" in kwargs:
+            t = type(kwargs["default"])
+            if t is bool:
+                return env == "True"
+            return t(env)
+        return env
+    except KeyError:
+        if "default" in kwargs:
+            return kwargs["default"]
+        raise RuntimeError(
```
Comment (mikerkelly) [inline opencodelists/settings.py:62]: Arguably `KeyError` is a better fit than `RuntimeError` if something is missing from the `os.environ` dict. I don't feel strongly though.

**Review by mikerkelly — COMMENTED** [review]
Could you please talk me through the `docker/docker-compose.yaml` change a bit more? I think in production we aren't using a `.env` file or `just`  -- those are dev tools. Is that right? And if so, does changing this docker-related file affect production at all? I don't think we should be doing that. I think in production the environment variables are configured by a different mechanism -- `dokku config:set opencodelists` etc. in `DEPLOY.md`.

I notice the second commit ([6da1873](https://github.com/opensafely-core/opencodelists/pull/2271/commits/4dceb7f62f418bb455f131fbe2d95e30ee8ad6ce)) doesn't pass all the checks. Somehow [the hint is being displayed when run in CI](https://github.com/opensafely-core/opencodelists/actions/runs/12517138207/job/34917415578#logs). Should this be happening? I'm not sure I understood why the other commits pass the checks though. I think if a repo with suitable env configuration this wouldn't happen so maybe it doesn't matter too much. Could some commits be merged to be atomic?

`get_env_var` seems to be doing three things: a parallel implementation of `os.getenv`; coercing `"True"` to `True`; and displaying the hint if a value is missing with no provided default.

I'm not sure handling defaults etc. ourselves here is gaining us anything over using `os.getenv`.

I think the coercing doesn't belong with the others and combing them is hard to follow. I think the variables that depend on the string `"True"` ought to be explicit about that rather than doing that logic in this function whose name doesn't indicate that.

I think the code as it is now is complicated enough that it ought to have a docstring and some tests. Maybe we can avoid that by simplifying a bit, something like (not tried):

```python
def get_env_var(key, default=None):
    env_var_value = os.getenv(key)
    if not env_var_value:
	raise KeyError(f"Missing environment variable: {key}\n\n{_missing_env_var_hint}")
    return env_var_value

...

DEBUG = get_env_var("DEBUG", default=False) == "True"
DEBUG_TOOLBAR = get_env_var("DJANGO_DEBUG_TOOLBAR", default=False) == "True"
IN_PRODUCTION = get_env_var("IN_PRODUCTION", default=False) == "True"
ASSETS_DEV_MODE = get_env_var("ASSETS_DEV_MODE", default=False) == "True"
```

**Comment by Jongmassey** [comment]
>I notice the second commit ([6da1873](https://github.com/opensafely-core/opencodelists/pull/2271/commits/4dceb7f62f418bb455f131fbe2d95e30ee8ad6ce)) doesn't pass all the checks. Somehow [the hint is being displayed when run in CI](https://github.com/opensafely-core/opencodelists/actions/runs/12517138207/job/34917415578#logs). Should this be happening? I'm not sure I understood why the other commits pass the checks though. 

Yep, this relates to an unexpected source of pain which I though was related to a cryptic message in a code comment in airlock about order of precedence of env vars when running a docker container; _however_ it wasn't actually that but was in fact the case that until the third commit in this PR, the env vars were being loaded from `.env` when running in a docker container.  The change to `docker-compose.yaml` in  a724d3b70cf4f2737fe980f93e61116a7f15bdde  fixes this. This previously wasn't necessary  as `environs` did this loading. Airlock [has similar configuration](https://github.com/opensafely-core/airlock/blob/6f747f192df94d75e551a2e290fe81edf3aaec11/docker/docker-compose.yml#L33) in its docker-compose

File: opencodelists/settings.py (hunk 2 above)
Comment (Jongmassey) [reply opencodelists/settings.py:62]: I don't have feelings either way, this was lifted straight from the airlock code

**Comment by Jongmassey** [comment]
>Could you please talk me through the docker/docker-compose.yaml change a bit more? I think in production we aren't using a .env file or just -- those are dev tools. Is that right? And if so, does changing this docker-related file affect production at all? I don't think we should be doing that. 

Right...

>I think in production the environment variables are configured by a different mechanism -- dokku config:set opencodelists etc. in DEPLOY.md.

They are, but also prior to this change, `opencodelists/settings.py` [uses](https://github.com/opensafely-core/opencodelists/blob/003f6941335710a1db3b3b218b001bcc9330fcc7/opencodelists/settings.py#L34-L35) environs to load these out of the `.env` file, this happens regardless of whether run in prod or dev; directly or via Docker. There isn't a `.env` file in the repo, but instead a [`dotenv-sample`](https://github.com/opensafely-core/opencodelists/blob/main/dotenv-sample) which is copied by the `_env` `just` recipe to a `.env` file. Since this recipe isn't called in prod, there isn't an `.env` file present in prod, and thus this configuration won't many any difference.

```sh
Jongmassey@dokku3:~$ dokku enter opencodelists
groups: cannot find name for group ID 10003
I have no name!@28bd6776c3df:/app$ ls .env
ls: cannot access '.env': No such file or directory
```

There are also some env vars [defined](https://github.com/opensafely-core/opencodelists/blob/003f6941335710a1db3b3b218b001bcc9330fcc7/docker/docker-compose.yaml#L33-L37) in the `prod` service in `docker-compose.yaml` but these are dummy values just used when checking the prod image builds correctly. They are overwritten by the `dokku config:set` operations you mention. Even if, for some reason, a `.env` file made it to production, these would also be overridden similarly.


So, since the pre-existing configuration supported provision of env vars for CI in docker from a `.env` file via `environs` removing `environs` means we've got to replace it with something else, in this case by specifying the `.env` file in `docker-compose.yaml`, as does [airlock](https://github.com/opensafely-core/airlock/blob/6148179ac2b0c669b0388fbaf2ce31ddeeaa3324/docker/docker-compose.yml#L33).

**Comment by Jongmassey** [comment]
>get_env_var seems to be doing three things: a parallel implementation of os.getenv; coercing "True" to True; and displaying the hint if a value is missing with no provided default.
>
>I'm not sure handling defaults etc. ourselves here is gaining us anything over using os.getenv.
>
>I think the coercing doesn't belong with the others and the combined logic is harder to follow. I think the variables that depend on the string "True" ought to be explicit about that rather than doing that logic in this function whose name doesn't indicate that.

The "parallel implementation of os.getenv" and "hinting" aspects are taken from [the airlock implementation](https://github.com/opensafely-core/airlock/blob/6f747f192df94d75e551a2e290fe81edf3aaec11/airlock/settings.py#L39-L45). However, I found a couple of things odd in airlock's `settings.py`, mostly to do with defaults and type coercion:

Despite having a function called `get_env_var` it also calls `os.environ.get` directly where a default value is required. It seemed seemed sensible to me to roll that default-provision behaviour into `get_env_var`, although acknowledge that in those cases it will never hit the missing value case.

It also does the `"True"`=>`True` dance in [a couple of places](https://github.com/opensafely-core/airlock/blob/6f747f192df94d75e551a2e290fe81edf3aaec11/airlock/settings.py#L63-L66) and so also it seemed sensible to make `get_env_var` the one place for getting env vars in the desired destination type. In my testing, I found that the variables previously got by `env.bool()` really did need to be bools and not strings and that changing the things that consume them to be happy with strings rather than bools would be impractical.


In summary, it's more of an attempt to be a drop-in replacement for `env.[something]()` and thus has some more complexity than just calling `os.getenv()`

**Comment by mikerkelly** [comment]
Thanks for the explanation of the CI fail + docker details, Jon.

I think before this ticket in prod the `read_env` call would have found an `.env` file if present, but it wouldn't have overridden any pre-existing env vars (those coming from dokku's configuration). Now, I think the prod app won't try to access `.env` at all, which seems even better. Applications should take the environment they're given. I'd not properly appreciated that the `docker-compose.yaml` `prod` service is only used for building the prod image, nothing to do with running it. So that's all good.

**Comment by mikerkelly** [comment]
On the approach to the getting the env vars in settings.py I suppose I'm saying I still prefer the Airlock approach. But that isn't a very strongly held opinion and if you disagree I'd be keen to understand why.

I think functions should ideally do one thing as concisely as possible and therefore be easy to read and understand through good naming, structure, and minimal-but-good docs. I don't see that a multi-function "one-stop-shop" function for "all of the env var accesses in this module" is better than the specific cases that want the hint, default, or parsing behaviour invoking those in each place that they're needed. That seems more concise, precise, explicit, and readable, to me.

> The "parallel implementation of os.getenv" and "hinting" aspects are taken from AIrlock

I saw that that's where the hinting idea comes from and I can marginally see the dev experience benefit to providing a hint where there is no default (although I would like the function name or docs to indicate it's only useful where there is no default, as that's not so obvious). (Marginal against the cost of baking in some assumptions about the application's possible environments, which may end up out-of-date or just wrong if copied somewhere else.)

The reference to `os.getenv` I made was around the default-provisioning behaviour that the Airlock implementation doesn't do -- it just wraps `os.environ[name]`.

> Despite having a function called `get_env_var` it also calls `os.environ.get` directly where a default value is required. It seemed seemed sensible to me to roll that default-provision behaviour into `get_env_var`, although acknowledge that in those cases it will never hit the missing value case.

I don't think I understand the intended benefit to writing our own default-provision behaviour when either `os.environ.get` or `os.getenv` already handle those cases well. It complicates the function and is one more thing to understand when reading the file and one more thing to maintain.

> It also does the `"True"`=>`True` dance in [a couple of places](https://github.com/opensafely-core/airlock/blob/6f747f192df94d75e551a2e290fe81edf3aaec11/airlock/settings.py#L63-L66) and so also it seemed sensible to make `get_env_var` the one place for getting env vars in the desired destination type.

I think it's better to do that conversion explicitly and directly where those bool variables are set than indirectly in `get_env_var` that is also doing other things.

I don't think there's a problem with writing that conversion in multiple places in a settings file (which is not much like reusable code components). A single-purpose function to do the conversion would be reasonable but I think it's marginal whether that would be more or less readable. But mixing it into a multi-purpose function doesn't seem like a good idea to me.

>In my testing, I found that the variables previously got by `env.bool()` really did need to be bools and not strings and that changing the things that consume them to be happy with strings rather than bools would be impractical.

That seems right to me.

**Comment by Jongmassey** [comment]
I've rewritten the changes to `settings.py` such that `get_env_var()` doesn't do any defaulting or type casting and instead rely on `os.environs.get()` and some inline casting/comparisons. Having done this, there are only two instances remaining where it is actually called but given it proved useful in finding the issue with running in docker I still think it merits inclusion.

Care to re-look @mikerkelly  ?

**Review by mikerkelly — APPROVED** [review]
Thanks, Jon. Looks good to me.
